packages/program/src/script-request.ts

```diff
@@ -0,0 +1,183 @@
+import type { BytesLike } from '@ethersproject/bytes';
+import { arrayify } from '@ethersproject/bytes';
+import { Logger } from '@ethersproject/logger';
+import {
+  VM_TX_MEMORY,
+  TRANSACTION_SCRIPT_FIXED_SIZE,
+  ASSET_ID_LEN,
+  WORD_SIZE,
+  CONTRACT_ID_LEN,
+} from '@fuel-ts/abi-coder';
+import type { BN } from '@fuel-ts/math';
+import type {
+  CallResult,
+  TransactionResultReceipt,
+  TransactionResultReturnReceipt,
+  TransactionResultReturnDataReceipt,
+  TransactionResultRevertReceipt,
+  TransactionResultScriptResultReceipt,
+  TransactionResult,
+} from '@fuel-ts/providers';
+import { ReceiptType, ByteArrayCoder } from '@fuel-ts/transactions';
+import { versions } from '@fuel-ts/versions';
+
+import { ScriptResultDecoderError } from './errors';
+import type { CallConfig } from './types';
+
+const logger = new Logger(versions.FUELS);
+
+export type ScriptResult = {
+  code: BN;
+  gasUsed: BN;
+  receipts: TransactionResultReceipt[];
+  scriptResultReceipt: TransactionResultScriptResultReceipt;
+  returnReceipt:
+    | TransactionResultReturnReceipt
+    | TransactionResultReturnDataReceipt
+    | TransactionResultRevertReceipt;
+  callResult: CallResult;
+};
+
+function callResultToScriptResult(callResult: CallResult): ScriptResult {
+  const receipts = [...callResult.receipts];
+
+  // Every script call ends with two specific receipts
+  // Here we check them so `this.scriptResultDecoder` doesn't have to
+  const scriptResultReceipt = receipts.pop();
```

Comment:
I know this is not coming from this PR, but this refactor just brought this to my attention now: we're hard-coding the assumptions that the receipts are ordered. If that changes in the future, it'll trigger some gnarly errors. 

I'm sure it's fine for now, but we should, maybe, refactor this not to assume order and filter the receipts we're looking for in the `receipts`. 

It shouldn't be a gating thing; we can open an issue for this, incur the tech debt now, and address this later.

Reply:
Issue created:
 - https://github.com/FuelLabs/fuels-ts/issues/802